core/src/code_hash.rs

```diff
@@ -0,0 +1,80 @@
+// This file is part of Gear.
+
+// Copyright (C) 2021-2022 Gear Technologies Inc.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Module for code hashes.
+
+use crate::checked_code::CheckedCode;
+use codec::{Decode, Encode};
+use core::convert::TryFrom;
+
+const HASH_LENGTH: usize = 32;
+
+/// Blake2b hash of the program code.
+#[derive(Clone, Copy, Debug, Decode, Encode, Ord, PartialOrd, Eq, PartialEq)]
+pub struct CodeHash([u8; HASH_LENGTH]);
+
+impl CodeHash {
+    /// Instantiates [`CodeHash`] by computing blake2b hash of the program `code`.
+    pub fn generate(code: &[u8]) -> Self {
+        let blake2b_hash = blake2_rfc::blake2b::blake2b(HASH_LENGTH, &[], code);
+        Self::from_slice_unchecked(blake2b_hash.as_bytes())
+            .expect("Blake2bResult has the specified len")
+    }
+
+    /// Get inner (32 bytes) array representation
+    pub fn inner(&self) -> [u8; HASH_LENGTH] {
+        self.0
+    }
+
+    /// Create new `CodeHash` bytes. Returns `None` if the provided slice has wrong length.
+    pub fn from_slice_unchecked(s: &[u8]) -> Option<Self> {
+        <[u8; HASH_LENGTH]>::try_from(s).map(|a| a.into()).ok()
+    }
+}
+
+impl From<[u8; HASH_LENGTH]> for CodeHash {
+    fn from(data: [u8; HASH_LENGTH]) -> Self {
+        CodeHash(data)
+    }
+}
+
+/// Contains checked code for a program and the hash for it.
+pub struct CheckedCodeHash(CheckedCode, CodeHash);
```

Comment:
Maybe the name `CheckedCodeWithHash` would be more clear? Also, I'm not excited with the tuple struct but it's ok for two fields.

Reply:
Sounds reasonable, will change